laurent22/joplin
# Desktop: Fixes #6054: Copy/Pasted URLs are Pasted in Plain Text [WYSIWYG editor]

# Overview
When pasting a link (or any URI) to the WYSIWYG editor, it is pasted as plaintext instead of a clickable link. However, when switching away from the note and back (or changing the editing mode) the plaintext link magically turns into a clickable link.

Here's what happens when the user switches away from the note after pasting a link:
![Switching Away](https://user-images.githubusercontent.com/36959928/151696895-df3786e7-a7ad-4d32-b792-c968484772e9.gif)

Here's what happens when the user changes the editing mode instead:
![Changing Mode](https://user-images.githubusercontent.com/36959928/151696921-3c1f130a-1f96-4263-a653-29fc0e7ecce2.gif)

I believe that the reason behind the bug is because the paste event is not being handled properly. When a link is being pasted, the event handler should detect that, convert the link into the markdown format, and then paste it.

# Problem Description
Discussion related to this particular issue has taken place in #5576, and about a similar issue in #5534. The key issue was described succinctly [here](https://github.com/laurent22/joplin/issues/5576#issuecomment-945061466):
> Switching between notes should not be changing the formatting of a note.

And [here](https://github.com/laurent22/joplin/issues/5534#issuecomment-950029356):

> > The links are clickable in the preview pane. I don't think that this is a bug.
>
> The whole point of using the WYSIWYG editor is that you don't have to use any other panels though.

# Solution
## User Experience
The user interface remains unchanged, but now instead of the previous behavior:
![Original Behavior](https://user-images.githubusercontent.com/36959928/151698148-7ae86d38-e1aa-4369-a77d-9677017f670a.gif)

Pasting links no longer pastes them as plaintext:

![Fixed Behavior](https://user-images.githubusercontent.com/36959928/151698187-9e4e8299-1205-472c-91d2-1a5bdf3c22e1.gif)

## Technical Solution
Initially in #6073, I had used a small external package to solve the issue. I switched it out based on laurent22's advice to rely on the `urlProtocol` function in the existing `@joplin/lib/urlUtils` library. This fixed the issue. I refrained from using the `valid-url` package (which is used in `app-mobile`) since it:
- Only has [196 stars](https://github.com/ogt/valid-url).
- Hasn't been updated in [8 years](https://www.npmjs.com/package/valid-url).
- Not used in `app-desktop`.

Following this I realized that the `urlUtils.urlProtocol(pastedText) !== null` check, while technically correct, was ambiguous and hard to test. Furthermore, it incorrectly identified `'http: // this is a broken link'` as a URL. So I switched to the latest [Node.js WHATWG URL API](https://nodejs.org/dist/latest-v8.x/docs/api/url.html#url_class_url) instead.

Then, I abstracted the logic and created a new function to test if a given string is a URI, named `isUri`, in the `urlUtils` module, and then converted the module to TypeScript as per [CONTRIBUTING.md](https://github.com/laurent22/joplin/blob/dev/CONTRIBUTING.md#coding-style).

# Testing Plan
After this, I added a test for my the function with 11 cases. The test cases test for `http://`, `https://` and `file:///` protocols (but other protocols are also supported), along with a very long link, a broken one, and a link containing Cyrillic characters.
## Steps
From the root of the project:
```
cd packages/lib
yarn test urlUtils.test
```

## Unit Test Results
![Test Result](https://user-images.githubusercontent.com/36959928/151698699-8d91aa68-f71f-4817-b326-984a062ea192.png)

## Manual Testing
I've conducted the following manual tests in an attempt to break the application or find bugs:
- ✅ Pasted a URL, closed the instance, and restarted it to check if it still renders as a link.
- ✅ Pasted a URL and switched between editing modes multiple times.
- ✅ Pasted a pre-formatted markdown URL `(https://google.com)[https://google.com]`, then switched between editing modes.
- ✅ Pasted a very long link to see if there was any visual lag. There was none.
- ✅ Pasted the link with several formatting modes (bold, italics, underline, list-item, etc).

I encourage others to conduct the manual tests as well. The build instructions are present [here](https://github.com/laurent22/joplin/blob/dev/BUILD.md).

# Additional Notes
If merged, *this will be my first code contribution to Joplin.* I apologize for any mistakes that I may have made in the previous PR, I'm always open to advice and guidance. I hope this fix works out, and I'm looking forward to contributing more to the codebase!
